test/protocol/integration/exchange/zeroExApiAdapter.spec.ts

```diff
@@ -646,4 +644,402 @@ describe("ZeroExApiAdapter", () => {
+          ETH_ADDRESS,
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched input token quantity");
+      });
+
+      it("rejects wrong output token quantity", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("sellTokenForEthToUniswapV3", [
+          encodePath([sourceToken, ETH_ADDRESS]),
+          sourceQuantity,
+          otherQuantity,
+          destination,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          ETH_ADDRESS,
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched output token quantity");
+      });
+
+      it("rejects invalid uniswap path", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("sellTokenForEthToUniswapV3", [
+          encodePath([sourceToken]),
+          sourceQuantity,
+          minDestinationQuantity,
+          destination,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          ETH_ADDRESS,
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("UniswapV3 token path too short");
+      });
+
+      it("rejects wrong destination", async () => {
+        const data = zeroExMock.interface.encodeFunctionData("sellTokenForEthToUniswapV3", [
+          encodePath([sourceToken, ETH_ADDRESS]),
+          sourceQuantity,
+          minDestinationQuantity,
+          ADDRESS_ZERO,
+        ]);
+        const tx = zeroExApiAdapter.getTradeCalldata(
+          sourceToken,
+          ETH_ADDRESS,
+          destination,
+          sourceQuantity,
+          minDestinationQuantity,
+          data,
+        );
+        await expect(tx).to.be.revertedWith("Mismatched recipient");
+      });
+    });
+
+    describe("sellEthForTokenToUniswapV3", () => {
+      const additionalHops = [otherToken, extraHopToken];
+      for (let i = 0; i <= additionalHops.length; i++) {
+        const hops = take(additionalHops, i);
+        it(`validates data for ${i + 1} hops`, async () => {
+          const path = [ETH_ADDRESS, ...hops, destToken];
+
+          const data = zeroExMock.interface.encodeFunctionData("sellEthForTokenToUniswapV3", [
+            encodePath(path),
+            minDestinationQuantity,
+            destination,
+          ]);
+          const [target, value, _data] = await zeroExApiAdapter.getTradeCalldata(
+            ETH_ADDRESS,
+            destToken,
+            destination,
+            ZERO,
+            minDestinationQuantity,
+            data,
+          );
+          expect(target).to.eq(zeroExMock.address);
+          // TODO(kimpers): is value 0 correct here?
+          expect(value).to.deep.eq(ZERO);
```

Comment:
Should value really be 0 here?

Reply:
I believe it should == `sourceQuantity` with the changes I suggested.